src/bundle-manifest.ts

```diff
@@ -219,15 +220,123 @@ export function mergeBundles(bundles: Bundle[]): Bundle {
 /**
  * A simple function for generating shared bundle names based on a counter.
  */
-export function sharedBundleUrlMapper(bundles: Bundle[]): UrlString[] {
+export function sharedBundleUrlMapper(bundles: Bundle[]):
+    Map<UrlString, Bundle> {
   let counter = 0;
-  const urls: UrlString[] = [];
+  const urlMap = new Map<UrlString, Bundle>();
   for (const bundle of bundles) {
     if (bundle.entrypoints.size === 1) {
-      urls.push(Array.from(bundle.entrypoints)[0]);
+      urlMap.set(Array.from(bundle.entrypoints)[0], bundle);
     } else {
-      urls.push(`shared_bundle_${++counter}.html`);
+      urlMap.set(`shared_bundle_${++counter}.html`, bundle);
+    }
+  }
+  return urlMap;
+}
+
+/**
+ * Compute the intersection of all of the passed `sets`.
+ * See: https://en.wikipedia.org/wiki/Intersection_(set_theory)
+ */
+function setIntersection<T>(sets: Set<T>[]): Set<T> {
+  return sets.reduce((previous, current) => {
+    const reduced = new Set<T>();
+    for (const entry of previous) {
+      if (current.has(entry)) {
+        reduced.add(entry);
+      }
     }
+    return reduced;
+  });
+}
+
+/**
+ * Returns a list of the sets of entrypoints, for use in performing
+ * set intersection and subtraction on the sets.
+ */
+function getEntrypointSets(bundles: Bundle[]): Set<string>[] {
+  const list: Set<string>[] = [];
+  for (let bundle of bundles) {
+    list.push(bundle.entrypoints);
   }
-  return urls;
+  return list;
 }
+
+/**
+ * Attempts to generate shared bundle names by identifying unique Entrypoints,
+ * falling back on `sharedBundleUrlMapper` when unable to form a direct
+ * correspondence.
+ *
+ * example:
+ * const bundle1 = {Entrypoints: ['A']};
```

Comment:
Is this for the polymer core build use case? I think we determined that we could name bundles by looking for entrypoints in the _dependencies_ of a bundle.

Polymer core should cause a bundle list of something like:

(std): [std, a, b, c]
(std, mini): [mini, d, e, f]
(std, mini, micro): [micro, g, h, i]

As long are there aren't any dependency cycles between entrypoints, there will be only one entrypoint in the dependencies list, and you can simply name the bundle after that.